### PR TITLE
chore: arm linux should work

### PR DESCRIPTION
**Pull request by dzmitry-lahoda**
- fixed all system check - do not export non existing attrs (no system)
- made 0.8.23 default
- used official rfc formatter and formatter all files
- update list input as it changed its hash (bug fix in place was made by solc people - crazy)
- fixed flake attribute exports warnings
- make sure on recursion in overlay
- ported source build to 8 23 and started templating (seems many changes wil be neede as it goes later versions)
- made error to be soft landing

### CODE REVIEW

**Comment by dzmitry-lahoda** [comment]
not tested, but do not see reason why it will not work.

**Comment by dzmitry-lahoda** [comment]
Should just work:
```
Dzmitry Lahoda
:computer:  7 minutes ago
not sure what you mean, but with nix:
me do not care official builds, i care only official make files - if there are mention of arm - i can build it
fixes for any arms are simple if official make files support it, for example https://github.com/nix-community/ethereum.nix/pull/386/files#diff-54a35e40130461d913a507bdde67091d7b7dd0054afb3c05a066a73e71bbd6b3R25
solc see is very popular, its make MUST support arm linux (which is VM on Mac, and Docker on Mac is VM)
hence just adding solc.nix one line should work, i sent PR, person active. he will do it soon i guess
so this just should work https://github.com/hellwolf/solc.nix/pull/6
(edited)

[#6 chore: arm linux should work](https://github.com/hellwolf/solc.nix/pull/6)
<https://github.com/[hellwolf/solc.nix](https://github.com/hellwolf/solc.nix)|hellwolf/solc.nix>hellwolf/solc.nix | Today at 5:23 PM | Added by [GitHub](https://null-studios-group.slack.com/services/B05BXAGG3JB)


Dzmitry Lahoda
:computer:  [3 minutes ago](https://null-studios-group.slack.com/archives/C05QGV097CN/p1715791891008839?thread_ts=1715788406.619529&cid=C05QGV097CN)
the do arm on CI
image.png
 
image.png




Dzmitry Lahoda
:computer:  [3 minutes ago](https://null-studios-group.slack.com/archives/C05QGV097CN/p1715791928087489?thread_ts=1715788406.619529&cid=C05QGV097CN)
https://github.com/ethereum/solidity/blob/816b588a01e4156f64dcfad8ffbf3b4132d0281d/.circleci/osx_install_dependencies.sh#L109

[osx_install_dependencies.sh](https://github.com/ethereum/solidity/blob/816b588a01e4156f64dcfad8ffbf3b4132d0281d/.circleci/osx_install_dependencies.sh)
    # be aware that we are only building the arm version here, we don't build a universal binary.
<https://github.com/[ethereum/solidity](https://github.com/ethereum/solidity)|ethereum/solidity>ethereum/solidity | Added by [GitHub](https://null-studios-group.slack.com/services/B05BXAGG3JB)


Dzmitry Lahoda
:computer:  [3 minutes ago](https://null-studios-group.slack.com/archives/C05QGV097CN/p1715791939136529?thread_ts=1715788406.619529&cid=C05QGV097CN)
it should just work
```

**Comment by dzmitry-lahoda** [comment]
first step to check, so  without new systems I get: so it is broken
```
       error: NAR hash mismatch in input 'https://binaries.soliditylang.org/macosx-amd64/list.json?narHash=sha256-akcOahJE/QR0jCPvo4aBf0loJt9P8zvP2E7oCBCRIho%3D' (/nix/store/9hw6yp4zv7vb7yq0s5yh42rxnjasf1p9-source), expected 'sha256-akcOahJE/QR0jCPvo4aBf0loJt9P8zvP2E7oCBCRIho=', got 'sha256-f5kzXsfnxgad+1FxaGOlflIUVt8RPNBGS5wkqb63DVA='
bash-5.1$ 
```

**Comment by dzmitry-lahoda** [comment]
have run flake lock update input, start getting this one

```
         | ^
          512|

       … while calling anonymous lambda

         at /nix/store/6gxrfravwqh3mjfnqvxgrl7hqm9cf3j1-source/pkgs/build-support/fetchurl/default.nix:123:40:

          122|   hash_ =
          123|     if with lib.lists; length (filter (s: s != "") [ hash outputHash sha1 sha256 sha512 ]) > 1
             |                                        ^
          124|     then throw "multiple hashes passed to fetchurl" else

       error: attribute 'solc-macos-aarch64' missing

       at /nix/store/qbgh3r0jdysqp5x52ghhwxizwmkmm4rb-source/mk-solc-static.nix:51:16:

           50|       inherit url;
           51|       sha256 = solc_sha256.${solc-flavor};
             |                ^
```

**Comment by dzmitry-lahoda** [comment]
@hellwolf so it works, so I did not done template, if you will have time to template - would be awesome. 

templating C++ sources good examples are in poetry2nix

**Comment by hellwolf** [comment]
Thanks @dzmitry-lahoda !!

That's a lot of changes to review, give me some time.

What autoformatting did you use?  Going forward, I should setup the same.

File: .github/workflows/flakehub-publish-rolling.yml [hunk 1]
```diff
@@ -8,8 +8,11 @@ jobs:
       id-token: "write"
       contents: "read"
     steps:
-      - uses: "actions/checkout@v3"
+      - uses: "actions/checkout@v4"
       - uses: "DeterminateSystems/nix-installer-action@main"
+      - uses: "DeterminateSystems/magic-nix-cache-action@v6"
+      - run: |
+          nix flake check --print-build-logs --show-trace --no-build --debug --verbose --all-systems
```
Comment (hellwolf) [inline .github/workflows/flakehub-publish-rolling.yml:15]: great tip!

**Comment by hellwolf** [comment]
Hi,

Thanks again for the inspiration.

I merged the change I did myself https://github.com/hellwolf/solc.nix/pull/7, which includes your changes:

- [x]  fixed all system check - do not export non existing attrs (no system)
- [ ] made 0.8.23 default
- [x] used official rfc formatter and formatter all files
- [x] update list input as it changed its hash (bug fix in place was made by solc people - crazy)
- [ ] fixed flake attribute exports warnings
- [ ] make sure on recursion in overlay
- [ ] ported source build to 8 23 and started templating (seems many changes wil be neede as it goes later versions)
- [ ] made error to be soft landing (?)

I will further incorporate your changes and additional cleanups:

- More code comments to make contribution easier.
- Check how to include your expression for building from source.
- Fix the rest of the TODOs

**Comment by dzmitry-lahoda** [comment]
> Check how to include your expression for building from source.

This would be awesome. It will make ultimate solution when there is no binary, just try build it. i guess that is main point.

> made error to be soft landing 

this is about do not throw, but just do not generate attribute (so it will just not find attr in consumer).

>  make sure on recursion in overlay

I got recursion somehow with flake-parts consumer. So imho better do explicit parameters passing in instead of pkgs, look  cosmos.nix for example.

>  ported source build to 8 23 and started templating (seems many changes wil be neede as it goes later versions)

latest nixpkgs is 8.21, I ported it to 8.23. I see that after 8.23 there are more changes in cmake. so need some flex system for CMAKE FetchContent override. So can start from 8.21 and 8.23 difference easy.

What can I do to help making these in master sooner than later? :)

**Comment by hellwolf** [comment]
I fixed the infinite recursion issue in main directly. But I also have a different issue in my local NixOS config using it. I will be fixing it asap this week!

**Comment by hellwolf** [comment]
https://github.com/hellwolf/solc.nix/pull/9 merging:

- CI to test ubuntu-latest, macos-12/13/14.
- Make solc 0.8.23 default.

**Comment by hellwolf** [comment]
I also realized that a deeper refactoring to use list files from https://github.com/ethereum/solc-bin/ would be better than maintaining my own lists.

**Comment by hellwolf** [comment]
I will checkout the "ported source build" expression you made.

I am not sure I would want to maintain that in this project yet, to be honest.

**Comment by dzmitry-lahoda** [comment]
> I am not sure I would want to maintain that in this project yet, to be honest.

You mean source version build or solc.nix? 

For first, build for existing version will not fail, so maintenance only where source is plugged as fallback for binary, which is 2  lines with `//` and `callPackage`.  Build itself is static, so it will work all the time without maintenance. Our company will migrate to new version, will PR add newer versions of source (8 25 and 8 23 could be built from source).

If second, ask if nix-community wants to maintain it and move to nix-community org. So these people I am pretty sure are on source version of things. But as you will look `nixpkgs`, it actually does both, source for linux and binary for mac. So you  approach is good. nix-community as org can add more reviewers than only you too (technically problems of own project is that owners of single person repo are single).

**Comment by dzmitry-lahoda** [comment]
for example request for recently moved project 

https://github.com/orgs/nix-community/discussions/835

**Comment by dzmitry-lahoda** [comment]
@hellwolf anyway, when you decide stop, and ok with having from source fallback in place, i will send new PR for that

**Comment by hellwolf** [comment]
> @hellwolf anyway, when you decide stop, and ok with having from source fallback in place, i will send new PR for that

Thanks again! I an bit slow at the moment, will look into it later this week. 

I think I am leaning towards maintaining it indeed, though I wish a bit more clarity in nixpkgs community how to deal with  "core" v.s. "contrib" packages.

**Comment by dzmitry-lahoda** [comment]
I am not nix community :) , so for sure never seen multiversion packages in nixpkgs. All multi version I saw were in nix-community repos (or repos of some orgs like cosmos.nix for example). Nixpkgs seems tracks latest stable release compatible with all other core pacakges in nixpkgs release.

**Comment by hellwolf** [comment]
Is " ported source build to 8 23 and started templating (seems many changes wil be neede as it goes later versions)" the remaining thing to address in a new pull request?

I am happy to get something like this merged, my 2cents:  since this project includes all versions of solidity, would it be a good idea to version the solc build expression too? For example, `mk-solc-from-source-0.8.23`, and perhaps 0.8.24 works the same so that the 0.8.23 will be reused over time.

**Comment by dzmitry-lahoda** [comment]
thanks, will open new PR for source.

**Comment by dzmitry-lahoda** [comment]
no worries, i am still target to follow up on this, so i guess to optimize, will do it when out team decides to move to next version of solc.

**Comment by msooseth** [comment]
Hi @dzmitry-lahoda is there an update on the way? I'd love to have arm :)

**Comment by hellwolf** [comment]
fwiw, I should switch the manually maintained list to https://github.com/ethereum/solc-bin/ instead. And if we make any progress on arm solc builds, ideally, it should be an upstream PR there too.

**Comment by dzmitry-lahoda** [comment]
we had it there for a while https://github.com/n1xyz/n1.nix/blob/master/solc-0.8.26.nix .

it is little bit hard to collaborate in this repo, expectancies for nix flake repos are different from what we have :)

**Comment by hellwolf** [comment]
> it is little bit hard to collaborate in this repo, expectancies for nix flake repos are different from what we have :)

Oh, could you elaborate a bit on that one?

**Pull request closed without merging**